# Logistic Regression

### 입력변수가 하나인 Logistic Regression

#### 1. 학습 데이터 (Training Data) 준비

In [1]:
import numpy as np

In [13]:
x_data = np.array([2,4,6,8,10,12,14,16,18,20]).reshape(10,1)
t_data = np.array([0,0,0,0,0,0,1,1,1,1]).reshape(10,1)

#### 2. 임의의 직선 z=Wx+b 정의 (임의의 값으로 가중치 W, 바이어스 b 초기화)

In [3]:
W = np.random.rand(1,1)
b = np.random.rand(1)
print("W = ",W,", W.shape = ", W.shape, ", b = ", b, ", b.shape = ", b.shape)

W =  [[0.94224825]] , W.shape =  (1, 1) , b =  [0.57634249] , b.shape =  (1,)


#### 3. 손실함수 E(W,b) 정의

In [4]:
# 시그모이드 구현
def sigmoid(x):
    return 1/(1+np.exp(-x))

In [5]:
# 손실함수 구현
def loss_func(x,t):
    delta = 1e-7    #무한대 발산 방지
    
    z = np.dot(x,W)+b
    y = sigmoid(z)
    
    return -np.sum(t*np.log(y+delta)+(1-t)*np.log((1-y)+delta))

#### 4. 수치미분 numerical_derivative 및 utility 함수 정의

In [6]:
# 수치미분 구현
def numerical_derivative(f,x): # x에 변수가 여러개
    delta_x = 1e-4
    grad = np.zeros_like(x)
    # 주어진 어레이와 같은 형태와 타입을 갖는 0으로 채워진 어레이를 반환
    # grad는 계산된 수치미분 값을 저장할 변수
    it = np.nditer(x,flags=['multi_index'],op_flags=['readwrite'])
    #  op_flags를 이용하면 기본값은 읽기 전용이지만 읽기/쓰기 또는 
    # 쓰기 전용 모드로 설정할 수 있다
    # multi_index는 뭐지..?
    while not it.finished:
        idx = it.multi_index
        
        tmp_val = x[idx]
        # x의 원본값이 변화되는 것을 막고자 임시변수 tmp_val 이라는 곳에
        # 원본값을 저장해두고 있다
        x[idx] = float(tmp_val) + delta_x
        fx1 = f(x) # f(x+delta_X)
        
        x[idx]=tmp_val - delta_x
        fx2 = f(x) # f(x-delta_X)
        grad[idx] = (fx1 - fx2) / (2*delta_x)
        
        x[idx] = tmp_val
        it.iternext()
    
    return grad

In [17]:
# 손실함수 값 계산 함수
def error_val(x,t):
    delta = 1e-7
    
    z = np.dot(x,W)+b
    y = sigmoid(z)
    
    return -np.sum(t*np.log(y+delta)+(1-t)*np.log((1-y)+delta))

# 학습을 모두 마친 후, 임의의 데이터에 대해 미래 값 예측 함수
def predict(x):
    
    z=np.dot(x,W)+b
    y=sigmoid(z)
    
    if y>0.5:
        result = 1
    else:
        result = 0
    return y, result

#### 5. 학습율(learning rate) 초기화 및 손실함수가 최소가 될 때까지 W,b 업데이트

In [15]:
learning_rate = 1e-2 # 발산하는 경우, 1e-3, 1e-6 등으로 바꾸어서 실행

f = lambda x : loss_func(x_data,t_data)
print("Initial erorr value = ", error_val(x_data,t_data))
print("Initial W = ",W)
print("Initial b = ",b)
      
for step in range(80001):
      
      W -= learning_rate*numerical_derivative(f,W)
      b -= learning_rate*numerical_derivative(f,b)
      if(step % 400 == 0):
        print("step = ", step )
        print("error_value = ", error_val(x_data,t_data))
        print("W = ",W)
        print("b = ",b)


Initial erorr value =  43.112628937169184
Initial W =  [[0.94224825]]
Initial b =  [0.57634249]
step =  0
error_value =  25.521486733347253
W =  [[0.52644351]]
b =  [0.51898398]
step =  400
error_value =  2.7567608183359447
W =  [[0.28365662]]
b =  [-4.15612679]
step =  800
error_value =  1.7753292006577623
W =  [[0.45549049]]
b =  [-5.66943644]
step =  1200
error_value =  1.513091737607316
W =  [[0.53232318]]
b =  [-6.69243868]
step =  1600
error_value =  1.3491726054817985
W =  [[0.59329399]]
b =  [-7.50203769]
step =  2000
error_value =  1.2335622579504824
W =  [[0.64462376]]
b =  [-8.18213601]
step =  2400
error_value =  1.145932521772633
W =  [[0.68939382]]
b =  [-8.77427257]
step =  2800
error_value =  1.0762570657988657
W =  [[0.72937221]]
b =  [-9.30226073]
step =  3200
error_value =  1.018939262912262
W =  [[0.76567493]]
b =  [-9.78111569]
step =  3600
error_value =  0.9705729672329803
W =  [[0.79905533]]
b =  [-10.22096198]
step =  4000
error_value =  0.9289481595832246
W =  

#### 6. 학습 결과 및 입력값을 통한 미래 값 예측

In [18]:
(real_val, logical_val) = predict(3)
print(real_val,logical_val)

[[2.72325993e-05]] 0


In [19]:
(real_val, logical_val) = predict(17)
print(real_val,logical_val)

[[0.98774741]] 1


 ### 입력변수가 여러개인 Logistic regression

#### 1. 학습 데이터 (Training Data) 준비

In [20]:
x_data = np.array([[2,4],[4,11],[6,6],[8,5],[10,7],[12,16],[14,8],[16,3],[18,7]])
t_data = np.array([0,0,0,0,1,1,1,1,1]).reshape(9,1)

#### 2. 임의의 직선 z=W1x1+W2x2+b 정의 (임의의 값으로 가중치 W, 바이어스 b 초기화)

In [21]:
W = np.random.rand(2,1)
b = np.random.rand(1)
print("W = ",W,", W.shape = ", W.shape, ", b = ", b, ", b.shape = ", b.shape)

W =  [[0.43400769]
 [0.03475751]] , W.shape =  (2, 1) , b =  [0.91019266] , b.shape =  (1,)


#### 3. 손실함수 E(W,b) 정의  - 동일

#### 4. 수치미분 numerical_derivative 및 utility 함수 정의 - 동일

#### 5. 학습율(learning rate) 초기화 및 손실함수가 최소가 될 때까지 W,b 업데이트 - 동일

In [23]:
learning_rate = 1e-2 # 발산하는 경우, 1e-3, 1e-6 등으로 바꾸어서 실행

f = lambda x : loss_func(x_data,t_data)
print("Initial erorr value = ", error_val(x_data,t_data))
print("Initial W = ",W)
print("Initial b = ",b)
      
for step in range(80001):
      
      W -= learning_rate*numerical_derivative(f,W)
      b -= learning_rate*numerical_derivative(f,b)
      if(step % 400 == 0):
        print("step = ", step )
        print("error_value = ", error_val(x_data,t_data))
        print("W = ",W)
        print("b = ",b)


Initial erorr value =  0.46535297730560965
Initial W =  [[1.19717424]
 [0.25282878]]
Initial b =  [-12.21371678]
step =  0
error_value =  0.4653181560950765
W =  [[1.19721923]
 [0.25285912]]
b =  [-12.2143054]
step =  400
error_value =  0.4518339595995166
W =  [[1.21484647]
 [0.26492666]]
b =  [-12.44594957]
step =  800
error_value =  0.43917270729244173
W =  [[1.23178593]
 [0.27684875]]
b =  [-12.67040504]
step =  1200
error_value =  0.4272505152204788
W =  [[1.24810411]
 [0.28861036]]
b =  [-12.88820278]
step =  1600
error_value =  0.41599595818449175
W =  [[1.263859  ]
 [0.30019877]]
b =  [-13.09980658]
step =  2000
error_value =  0.4053476951177657
W =  [[1.27910114]
 [0.31160373]]
b =  [-13.30562453]
step =  2400
error_value =  0.3952526251462699
W =  [[1.29387457]
 [0.3228174 ]]
b =  [-13.50601808]
step =  2800
error_value =  0.3856644418097265
W =  [[1.30821773]
 [0.3338342 ]]
b =  [-13.70130949]
step =  3200
error_value =  0.3765424891726206
W =  [[1.3221642 ]
 [0.34465058]]
b 

step =  32400
error_value =  0.1373441757215671
W =  [[1.90966624]
 [0.79711647]]
b =  [-21.88740408]
step =  32800
error_value =  0.13614590092084897
W =  [[1.91482518]
 [0.80088487]]
b =  [-21.95638334]
step =  33200
error_value =  0.13496812319833654
W =  [[1.91994107]
 [0.8046189 ]]
b =  [-22.02476978]
step =  33600
error_value =  0.13381032495792947
W =  [[1.9250146 ]
 [0.80831919]]
b =  [-22.09257339]
step =  34000
error_value =  0.13267200579939328
W =  [[1.93004647]
 [0.81198633]]
b =  [-22.1598039]
step =  34400
error_value =  0.13155268181224797
W =  [[1.93503734]
 [0.8156209 ]]
b =  [-22.22647082]
step =  34800
error_value =  0.13045188490423626
W =  [[1.93998787]
 [0.81922347]]
b =  [-22.29258341]
step =  35200
error_value =  0.12936916216220343
W =  [[1.9448987]
 [0.8227946]]
b =  [-22.35815072]
step =  35600
error_value =  0.12830407524372192
W =  [[1.94977044]
 [0.82633481]]
b =  [-22.42318156]
step =  36000
error_value =  0.12725619979768255
W =  [[1.95460371]
 [0.82984

step =  64400
error_value =  0.0804147442882717
W =  [[2.22596422]
 [1.02363672]]
b =  [-26.08999986]
step =  64800
error_value =  0.0799986551447914
W =  [[2.22903906]
 [1.02580179]]
b =  [-26.13063796]
step =  65200
error_value =  0.07958682305533307
W =  [[2.23209837]
 [1.02795537]]
b =  [-26.17106756]
step =  65600
error_value =  0.07917918331860471
W =  [[2.23514231]
 [1.03009756]]
b =  [-26.21129078]
step =  66000
error_value =  0.07877567253325218
W =  [[2.23817104]
 [1.03222849]]
b =  [-26.2513097]
step =  66400
error_value =  0.07837622856547505
W =  [[2.24118471]
 [1.03434827]]
b =  [-26.29112638]
step =  66800
error_value =  0.07798079051759182
W =  [[2.24418346]
 [1.03645702]]
b =  [-26.33074283]
step =  67200
error_value =  0.07758929869754998
W =  [[2.24716743]
 [1.03855485]]
b =  [-26.37016105]
step =  67600
error_value =  0.07720169458931288
W =  [[2.25013678]
 [1.04064188]]
b =  [-26.40938299]
step =  68000
error_value =  0.07681792082410527
W =  [[2.25309164]
 [1.0427

#### 7. 미래 값 예측

In [24]:
test_data = np.array([3,17])
predict(test_data)

(array([0.13923044]), 0)

In [25]:
test_data = np.array([5,8])
predict(test_data)

(array([0.00086224]), 0)

In [26]:
test_data = np.array([7,21])
predict(test_data)

(array([0.99999336]), 1)

In [27]:
test_data = np.array([12,0])
predict(test_data)

(array([0.62013865]), 1)

  ===> 미래 값 예측을 통해 복습보다는 예습시간이 합격에 미치는 영향이 크다는 것을 알 수 있었다. 예습시간에 대한 가중치 W1은 2.33, 복습시간에 대한 가중치 W2 는 1.1에서 보듯이 예습시간이 복습시간에 비해 최종결과에 미치는 영향이 2배 이상임을 알 수 있다